src/main/java/org/cactoos/io/LoggingInputStream.java

```diff
@@ -0,0 +1,236 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.cactoos.io;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Logged input stream.
+ *
+ * <p>There is no thread-safety guarantee.
+ *
+ * @author Fabricio Cabral (fabriciofx@gmail.com)
+ * @version $Id$
+ * @since 0.29
+ */
+@SuppressWarnings("PMD.LoggerIsNotStaticFinal")
+public final class LoggingInputStream extends InputStream {
+
+    /**
+     * The input stream.
+     */
+    private final InputStream origin;
+
+    /**
+     * Where the data comes from.
+     */
+    private final String source;
+
+    /**
+     * The logger.
+     */
+    private final Logger logger;
+
+    /**
+     * The bytes read.
+     */
+    private final AtomicLong bytes;
+
+    /**
+     * The time took to read.
+     */
+    private final AtomicLong time;
+
+    /**
+     * Ctor.
+     * @param input Source of data
+     * @param src The name of source data
+     */
+    public LoggingInputStream(final InputStream input, final String src) {
+        this(input, src, Logger.getLogger(src));
+    }
+
+    /**
+     * Ctor.
+     * @param input Source of data
+     * @param src The name of source data
+     * @param lgr The message logger
+     */
+    public LoggingInputStream(
+        final InputStream input,
+        final String src,
+        final Logger lgr
+    ) {
+        super();
+        this.origin = input;
+        this.source = src;
+        this.logger = lgr;
+        this.bytes = new AtomicLong();
+        this.time = new AtomicLong();
+    }
+
+    @Override
+    public int read() throws IOException {
+        final byte[] buf = new byte[1];
+        this.read(buf);
+        return Byte.toUnsignedInt(buf[0]);
+    }
+
+    @Override
+    public int read(final byte[] buf) throws IOException {
+        return this.read(buf, 0, buf.length);
+    }
+
+    @Override
+    public int read(final byte[] buf, final int offset, final int len)
+        throws IOException {
+        final Instant start = Instant.now();
+        final int byts = this.origin.read(buf, offset, len);
+        final Instant end = Instant.now();
+        final long millis = Duration.between(start, end).toMillis();
+        final Level level = this.logger.getLevel();
+        final String msg = "Read %d byte(s) from %s in %dms.";
+        if (byts > 0) {
+            this.bytes.getAndAdd(byts);
+            this.time.getAndAdd(millis);
+            if (!level.equals(Level.INFO)) {
+                this.logger.log(
+                    level,
+                    String.format(
```

Comment:
@fabriciofx Can't you build this String upwards (and get rid of most of the local variables which are now used in 2 places, but then would be used in only one place)?

Reply:
@amihaiemil Yeah, I'll change it. Thanks.